main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -1242,6 +1309,75 @@ object Lowering {
     mkTag(Enums.HeadTerm, s"App$arity", innerExp, Types.HeadTerm, loc)
   }
 
+  /**
+    * Make the array of MpmcAdmin objects which will be passed to `selectFrom`
+    */
+  private def mkChannelAdminArray(rs: List[SelectChannelRule], channels: Array[(Symbol.VarSym, Expression)], pur: Type, eff: Type, loc: SourceLocation): Expression = {
+    val admins = rs.zipWithIndex map {
+      case (SelectChannelRule(_, c, _), i) =>
+        val admin = Expression.Def(Defs.ChannelMpmcAdmin, Type.mkPureArrow(c.tpe, Types.ChannelMpmcAdmin, loc), loc)
+        Expression.Apply(admin, List(Expression.Var(channels(i)._1, c.tpe, loc)), Types.ChannelMpmcAdmin, pur, eff, loc)
+    }
+    mkArray(admins, Types.ChannelMpmcAdmin, loc)
+  }
+
+  /**
+    * Construct a call to `selectFrom` given an array of MpmcAdmin objects and optional default
+    */
+  private def mkChannelSelect(adminArray: Expression, default: Option[Expression], pur: Type, eff: Type, loc: SourceLocation): Expression = {
+    val locksType = Types.mkList(Types.ConcurrentReentrantLock, loc)
+
+    val selectRetTpe = Type.mkTuple(List(Type.Int32, locksType), loc)
+    val selectTpe = Type.mkImpureUncurriedArrow(List(adminArray.tpe, Type.Bool), selectRetTpe, loc)
+    val select = Expression.Def(Defs.ChannelSelectFrom, selectTpe, loc)
+    val blocking = default match {
+      case Some(_) => Expression.False(loc)
+      case None => Expression.True(loc)
+    }
+    Expression.Apply(select, List(adminArray, blocking), selectRetTpe, pur, eff, loc)
+  }
+
+  /**
+    * Construct a sequence of MatchRules corresponding to the given SelectChannelRules
+    */
+  private def mkChannelCases(rs: List[SelectChannelRule], channels: Array[(Symbol.VarSym, Expression)], retTpe: Type, pur: Type, eff: Type, loc: SourceLocation)(implicit flix: Flix): List[MatchRule] = {
+    val locksType = Types.mkList(Types.ConcurrentReentrantLock, loc)
+
+    rs.zipWithIndex map {
+      case (SelectChannelRule(sym, chan, exp), i) =>
+        val locksSym = mkLetSym("locks", loc)
+        val pat = Pattern.Tuple(List(Pattern.Int32(i, loc), Pattern.Var(locksSym, locksType, loc)), retTpe, loc)
+        val getTpe = chan.tpe match {
```

Comment:
@mlutze Is this safe here or not?

Reply:
I think we need to call `eraseTopAliases` here.

Reply:
Right. This is important otherwise things could go badly.

Reply:
Is there a test that I could write before making this change to confirm that a) it's a problem as it currently stands and b) that I've fixed it?

Reply:
Yes, if you do something like:

type alias MyChannel = Channel[Int32]
def foo(c: MyChannel): ... /* use select on c)

Reply:
Hmm, so this passes. What am I missing?

```
    type alias MyChannel = Channel[Int32]

    @test
    def testSelectAliasedChannel(): Bool \ IO = {
        let c = chan Int32 1;

        def useChan(ch: MyChannel) : Bool =
            select {
                case x <- ch => x == 42
            };

        c <- 42;
        useChan(c)
    }
```

Reply:
Oh! But I am getting failures elsewhere (not sure why this wasn't happening before?)

Investigating...

Reply:
Ah! Because Magnus recently changed it to `Receiver` and `Sender`!